README
Status 404 Find Best Match Page

https://github.com/peterkahl/status-404-find-best-match-page



This PHP class catches all "soft errors 404" and finds the closest (best,
most likely) candidate page name (URI) from a list of pages that exist.

Soft error 404 is defined as one that is initiated by this PHP script:
* The script compares the current request against a list of existant pages.
* If match isn't found (soft error 404), a closest match is found.
* The script replies with status 302 (permanent redirect) to the existant
  page.

Hard error 404 is initiated by HTTP server.

Functionality of this class depends on properly set up rewrite inside
htaccess (or nginx configuration) file.

NGINX example:

server {
	error_page  404  /index.php?page=/404/;
	location / {
        rewrite ^(/[a-zA-Z0-9\-/\s]*)$   /index.php?page=$1   last;
    }
}

//----------------------------------------------------------------------



CHANGELOG
1.0.0 ..... 2013-08-26
	Initial release


example.php
<?php

// Status 404 Find Best Match Page
// Example of usage

// Sometimes, status 404 is triggered inside your HTTP server. In such case,
// the server is requesting page '/404/'. This is defined inside your
// htaccess file (or config file of nginx).
// Clearly, page '/404/' is one that exists.


// We *must* have an array that contains the valid URIs of all pages that exist.
// We should agree that only these characters can be used as valid page name:
//    a-z, 0-9, - and _

$exists = array(
	'/',
	'/about/',
	'/products/',
	'/products/antique-cars/',
	'/contact/',
	'/404/'
	);

// Requested page name

$request = reset(explode('?', $_SERVER["REQUEST_URI"]));

// Let's see if requested page exists!

//----------------------------------------------------------------------
if (!in_array($request, $exists)) {
	// requested page doesn't exist
	require dirname(__FILE__).'/class.status-404-find-best-match-page.php';
	$correction = new status_404_find_best_match_page($request, $exists);
	$correction->best_match(); // Here we will redirect with status 302 to a page with closest match URI.
}
//----------------------------------------------------------------------

// If we got this far, the requested page exists.

// And below put more of your code that generates the web page...

if ($request == '/') {
	echo 'Home';
}
elseif ($request == '/about/') {
	echo 'About';
}
elseif ($request == '/products/') {
	echo 'Products';
}
elseif ($request == '/products/antique-cars/') {
	echo 'Products > Antique Cars';
}
elseif ($request == '/contact/') {
	echo 'Contact';
}
elseif ($request == '/404/') {
	echo 'Status 404: Page not found';
	header("HTTP/1.1 404 Not Found");
}
else {
	echo 'ERROR: You have an error in your code.';
}

//----------------------------------------------------------------------




class.status-404-find-best-match-page.php
<?php

/**
 *
 * Status 404 Find Best Match Page
 * Copyright (c) 2013 Peter Kahl. All rights reserved.
 * Use of this source code is governed by a GNU General Public License
 * that can be found in the LICENSE file.
 * Version 1.0.0 ..... 2013-08-26
 *
 * https://github.com/peterkahl/status-404-find-best-match-page
 *
 */


class status_404_find_best_match_page {

	private $exists;
	private $request;

	// If unable to find closest match, default to:
	// (change as needed)
	private $closest = '/';

	public function __construct($request, $exists, $qstr = '') {
		$this->exists  = $exists;
		$this->request = $request;
		if (empty($qstr)) $this->qstr = '';
		else $this->qstr = '?'.$qstr;
	}

	public function best_match() {
		$a = str_replace(array('/','-','_','"',"'",'%20',' '), '', strtolower($this->request));
		$distance = -1;
		foreach ($this->exists as $pg) {
			if ($pg != '/404/') {
				$b = str_replace(array('/','-','_'), '', $pg);
				$lev = levenshtein($a, $b);
				// exact match
				if ($lev == 0) {
					$this->closest = $pg;
					$distance = 0;
					break;
				}
				elseif ($lev <= $distance || $distance < 0) {
					$this->closest = $pg;
					$distance = $lev;
				}
			}
		}
		$len1 = strlen($a);
		if (1.5*$distance > $len1) {
			foreach ($this->exists as $pg) {
				if ($pg != '/404/') {
					$b = str_replace(array('/','-','_'), '', $pg);
					$len2 = strlen($b);
					if ($len1 > 0 && $len2 > 0 && $len1 > $len2) {
						if (strstr($a, $b) !== false) {
							$this->closest = $pg;
							break;
						}
					}
					elseif ($len1 > 0 && $len2 > 0) {
						if (strstr($b, $a) !== false) {
							$this->closest = $pg;
							break;
						}
					}
				}
			}
		}
		// reload
		header("HTTP/1.1 301 Moved Permanently");
		header("Location: http://".$_SERVER["HTTP_HOST"].$this->closest.$this->qstr);
		exit();
	}
}

//----------------------------------------------------------------------



index.php
<?php

// Status 404 Find Best Match Page
// Example of usage

//-------
// Sometimes, status 404 is triggered inside your HTTP server. In such case,
// the server is requesting page '/404/'. This is defined inside your
// htaccess file (or config file of nginx).
// Clearly, page '/404/' is one that exists.
//-------
// This example assumes that inside your htaccess file (or config file of nginx)
// you rewrite all (qualifying) requests into index.php file.
// Example of nginx configuration:
//
//	error_page  404  /index.php?page=/404/;
//	location / {
//        rewrite ^(/[a-zA-Z0-9\-/\s]*)$   /index.php?page=$1   last;
//    }
//
//-------
// Only those pages '/404/' requested by your HTTP server will be considered as
// "hard error 404".
// Other pages will be considered as "soft error 404" (status 302) if they are 
// not listed in the valid URI array below.
//-------


// We *must* have an array that contains the valid URIs of all pages that exist.
// We should agree that only these characters can be used in a valid page name:
//    a-z, 0-9, - and _

$valid_uris = array(
	'/',
	'/about/',
	'/products/',
	'/products/antique-cars/',
	'/products/contemporary-cars/',
	'/products/kitchen-sink/',
	'/contact/',
	'/legal-terms-and-conditions/',
	'/404/'
	);

// NOTE: If you use other name for your page 404 other than '/404/', you'll
// need to change it inside the class.

// Requested page name, query string:

$request = $_GET["page"];

if (strstr($_SERVER["REQUEST_URI"], '?') !== false) {
	list($x, $qstr) = explode('?', $_SERVER["REQUEST_URI"]);
}
else {
	$qstr = '';
}

// Let's see if requested page exists!

//----------------------------------------------------------------------
if ($request != '/404/' && !in_array($request, $valid_uris)) {
	// The requested page doesn't exist.
	// Let's call this a "soft error 404" and we will reply with
	// status 302 instead (redirect to a page that exists).
	require dirname(__FILE__).'/class.status-404-find-best-match-page.php';
	$correction = new status_404_find_best_match_page($request, $valid_uris, $qstr);
	$correction->best_match(); // Here we will redirect with status 302 to a page with best match URI.
}
//----------------------------------------------------------------------

// If we got this far, the requested page exists.

// And below put more of your code that generates the web page...
// For example...

if ($request == '/') {
	echo 'Home';
}
elseif ($request == '/about/') {
	echo 'About';
}
elseif ($request == '/products/') {
	echo 'Products';
}
elseif ($request == '/products/antique-cars/') {
	echo 'Products > Antique Cars';
}
elseif ($request == '/products/contemporary-cars/') {
	echo 'Products > Contemporary Cars';
}
elseif ($request == '/products/kitchen-sink/') {
	echo 'Products > Kitchen Sink';
}
elseif ($request == '/contact/') {
	echo 'Contact';
}
elseif ($request == '/legal-terms-and-conditions/') {
	echo 'Legal Terms and Conditions';
}
elseif ($request == '/404/') {
	echo 'Status 404: Page not found';
	header("HTTP/1.1 404 Not Found");
}
else {
	echo 'ERROR: You have an error in your code.';
}

//----------------------------------------------------------------------


